src/request_executor.cc

```
@@ -56,12 +56,110 @@ void
+        }
+      }
+      std::shared_ptr<PbError> pb_error;
+      infer_response =
+          std::make_unique<InferResponse>(output_tensors, pb_error);
+    }
+    catch (const PythonBackendException& pb_exception) {
+      if (response != nullptr) {
+        LOG_IF_ERROR(
+            TRITONSERVER_InferenceResponseDelete(response),
+            "Failed to delete inference response.");
+
+        response = nullptr;
+      }
+      std::shared_ptr<PbError> pb_error =
+          std::make_shared<PbError>(pb_exception.what());
+      infer_response = std::make_unique<InferResponse>(
+          std::vector<std::shared_ptr<PbTensor>>{}, pb_error);
+    }
+
+    infer_response->SetCompletedResponse(response);
+
+    if (!(*p)->IsDecoupled()) {
+      infer_response->ResetNextResponseFuture();
+      (*p)->SetValueForPrevPromise(std::move(infer_response));
+      (*p)->ResetPrevPromise();
+    } else {
+      if ((flags & TRITONSERVER_RESPONSE_COMPLETE_FINAL) == 0) {
+        // Not the last reponse. Need to store the promise associated with the
+        // next future.
+        auto promise = new std::promise<std::unique_ptr<InferResponse>>();
+        infer_response->SetNextResponseFuture(promise);
+        (*p)->SetValueForPrevPromise(std::move(infer_response));
+        (*p)->SetPrevPromise(&promise);
+      } else {
+        // The last response.
+        infer_response->ResetNextResponseFuture();
+        (*p)->SetValueForPrevPromise(std::move(infer_response));
+        (*p)->ResetPrevPromise();
+      }
+    }
+  } else if (
+      (*p)->IsDecoupled() &&
+      (flags & TRITONSERVER_RESPONSE_COMPLETE_FINAL) != 0) {
+    // An empty response may be the last reponse for decoupled models.
+    (*p)->SetValueForPrevPromise(std::unique_ptr<InferResponse>{});
+    (*p)->ResetPrevPromise();
+  } else {
+    (*p)->SetValueForPrevPromise(std::unique_ptr<InferResponse>{});
+    (*p)->ResetPrevPromise();
+    throw PythonBackendException("Unexpected empty response.");
```

I don't think you can throw an exception here.

---

Great catch. Updated to make the error being returned as a part of  `InferResponse` object rather than throwing exception here.